File: frontend/src/lib/admin/audit.ts
import { apiFetch } from "@/lib/api/client";

export interface AuditActor {
  id: string;
  full_name: string | null;
  email: string;
}

export interface AdminAction {
  id: string;
  actor: AuditActor | null;
  action: string;
  target_type: string;
  target_id: string | null;
  action_metadata: Record<string, unknown>;
  created_at: string;
}

export interface AuditLogResponse {
  items: AdminAction[];
  total: number;
  next_offset: number | null;
}

export function getAuditLog(params: {
  action?: string;
  target_type?: string;
  actor_id?: string;
  limit?: number;
  offset?: number;
}) {
  const qs = new URLSearchParams();
  if (params.action) qs.set("action", params.action);
  if (params.target_type) qs.set("target_type", params.target_type);
  if (params.actor_id) qs.set("actor_id", params.actor_id);
  qs.set("limit", String(params.limit ?? 50));
  qs.set("offset", String(params.offset ?? 0));
  return apiFetch<AuditLogResponse>(`/admin/audit?${qs.toString()}`);
}

File: backend/app/schemas/audit.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class AuditActor(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    full_name: str | None = None
    email: str


class AdminActionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    actor: AuditActor | None
    action: str
    target_type: str
    target_id: str | None
    action_metadata: dict
    created_at: datetime


class AuditLogResponse(BaseModel):
    items: list[AdminActionRead]
    total: int
    next_offset: int | None = None
